worksheets/espresso-testing/index.md

```diff
@@ -0,0 +1,240 @@
+We should now do something a little more interesting with our tests. We are going to test that pressing on the cookie increases our total counter.
+
+Create a new method called `totalIncreasesWhenCookieClicked`. Inside this method add the following code:
+
+```java
+onView(withId(R.id.imgCookie))
+        .perform(click());
+```
+
+This looks similar to the code before, except this time we are performing an action rather than checking something. The first line finds the view with the id `imgCookie` (take a look at your `activity-main.xml` file if you can't remember which view this was). The second line then performs a click action on it. This will call our `OnClickListener` in our `MainActivity.java` file and should update the total label. We now need to write a check for that.
+
+To do this, add in these lines to the new test method:
+
+```java
+onView(withId(R.id.lblTotal))
+        .check(matches(withText("1")));
+```
+
+This is almost exactly the same as the code we wrote for our `totalStartsAtZero` test, but we're instead checking to see that it is now `"1"`.
+
+Our completed test should now look like this:
+
+```java
+@Test
+public void totalIncreasesWhenCookieClicked() throws Exception {
+    onView(withId(R.id.imgCookie))
+            .perform(click());
+
+    onView(withId(R.id.lblTotal))
+            .check(matches(withText("1")));
+}
+```
+
+Let's run this new test using the play icon next to the `totalIncreasesWhenCookieClicked` method line. This test should take just long enough for you to see the 1 on the screen, but it's still so fast that you might blink and miss it!
+
+## Getting a high score!
+We've now tested all the functionality of our app, so now let's do something just for fun: let's make Espresso click loads of time on the cookie and get a high score!
+
+Create a new method called `achieveHighScore`. Copy the code the click test you just wrote and wrap the code to click the cookie image in a for loop from 0 to 100. You then need to check that the total counter has reached `"100"` after the for loop has completed. Ask your coach, if you're not sure how to do this, or take a peek at the code below:
+
+```java
+@Test
+public void achieveHighScore() throws Exception {
+    for (int i = 0; i < 100; i++) {
+        onView(withId(R.id.imgCookie))
+                .perform(click());
+    }
+
+    onView(withId(R.id.lblTotal))
+            .check(matches(withText("100")));
+}
+```
+
+If you run this test it should take long enough that you can see what's going on.
+
+As well as running the tests one by one, you can run all your apps tests by pressing on the "double play" icon next to the `public class CookieClickerTest` line. This will run all the tests in this file one by one and let you know which ones pass and fail.
+
+![All tests passing](images/all_tests_passing.png)
+
+## Further information
+You have now tested your cookie clicker app!
+
+You can read more about Android testing using the links below:
```

Comment:
really great informative tutorial (where were you 6 months ago when I learned espresso for the first time!!!!)

Maybe consider adding an extension task at the end? Something students could consider adding on their own? Maybe using a text field for their name (for a high score message) and checking if it's not empty? What do you think?